### PR TITLE
[Merged by Bors] - feat(library/data/buffer/parser): add parser.{any_char,digit,nat}

### PR DESCRIPTION
**Pull request by JLimperg**
Add three basic parsers:

- `any_char` consumes and returns any character
- `digit` parses a digit
- `nat` (naively) parses a natural number

### CODE REVIEW

File: library/data/buffer/parser.lean [hunk 1]
```diff
@@ -188,6 +199,22 @@ def fix_core (F : parser α → parser α) : ∀ (max_depth : ℕ), parser α
 | 0             := failure
 | (max_depth+1) := F (fix_core max_depth)
 
+/-- Matches a digit (0-9). -/
+def digit : parser nat := do
+  c ← any_char,
```
Comment (gebner) [inline library/data/buffer/parser.lean:204]:
```suggestion
  c ← sat (λ c, '0' ≤ c ∧ c ≤ '9'),
```

Then the error message will appear at the right position and backtracking will work as expected.

File: library/data/buffer/parser.lean [hunk 2]
```diff
@@ -188,6 +199,22 @@ def fix_core (F : parser α → parser α) : ∀ (max_depth : ℕ), parser α
...
+  let c' := c.to_nat - '0'.to_nat,
+  if 0 ≤ c' ∧ c' ≤ 9
+    then pure c'
+    else parser.fail $ "expected a digit, got: " ++ c.to_string
+
+/-- Matches a natural number. Large numbers may cause performance issues, so
+don't run this parser on untrusted input. -/
+def nat : parser nat := do
```
Comment (gebner) [inline library/data/buffer/parser.lean:212]:
```suggestion
def nat : parser nat := decorate_error "<number>" $ do
```
Better error-message and backtracking.  (The `<|>` only backtracks if the position doesn't advance.)

File: library/data/buffer/parser.lean (hunk 1 above)
Comment (JLimperg) [reply library/data/buffer/parser.lean]: Fixed, thanks!

File: library/data/buffer/parser.lean (hunk 2 above)
Comment (JLimperg) [reply library/data/buffer/parser.lean]: Added the `decorate_error`. I'm afraid I don't understand the `<|>` comment -- was there supposed to be another suggested change?

Comment (gebner) [reply library/data/buffer/parser.lean]: No this was just (an apparently failed attempt at) an explanation.

Comment (JLimperg) [reply library/data/buffer/parser.lean]: Ah okay, thanks!

**Pull request closed without merging**